### ISSUE TITLE
Average gradients before allreduce

### ISSUE THREAD

**Issue by szhengac** [horovod/horovod#1699]
When a large number of workers with half-precision are used, allreduce can possibly cause overflow. Thus, one solution is to divide the gradient by number of workers before performing allreduce while the averaging is performed after the allreduce in horovod.

**Comment by tgaddair** [comment]
Hey @szhengac, have you observed overflow issues during training?  We experimented with a divide-before-sum approach in the past, and found that it was more likely to lead to underflow (as gradients tended to have smaller magnitudes on average, see [this](https://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html) page from NVIDIA).

But we can definitely consider providing both options as a config if necessary.

**Comment by szhengac** [comment]
> Hey @szhengac, have you observed overflow issues during training? We experimented with a divide-before-sum approach in the past, and found that it was more likely to lead to underflow (as gradients tended to have smaller magnitudes on average, see [this](https://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html) page from NVIDIA).
> 
> But we can definitely consider providing both options as a config if necessary.

Yes. I observed overflow when I trained large bert with hundreds of gpus. Providing both options looks good to me.

**Comment by tgaddair** [comment]
Sounds reasonable, should be pretty straightforward.  We already have the branch containing the divide-before-sum implementation on GitHub [here](https://github.com/horovod/horovod/compare/fp16_divide_before_sum), though it's from an old version of Horovod.

**Comment by tgaddair** [comment]
cc @romerojosh

**Comment by tgaddair** [comment]
@szhengac, can you try out the `gradient_predivide_factor` introduced in v0.20.0 and let us know if it works for your scenario?

**Comment by szhengac** [comment]
I may try it when I have access to that many machines again someday :).
